### ISSUE TITLE
Overlay predict_drug_treats_disease doesn't like NodeSynonymizer's 'preferred curies'

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#880]
This relates to issue #823, and to changes in the branch `node-normalization` (which I branched off of `expander` before plugging the new `NodeSynonymizer` into Expand, in order to keep `expander` relatively stable.)

The following overlay tests are failing in the `node-normalization` branch. I'm guessing the issue relates to the fact that the 'preferred curies' obtained from the `NodeSynonymizer` may not be the ones `predict_drug_treats_disease` wants (e.g., the KG contains CHEBI curies, but it wants CHEMBL?). Perhaps Overlay needs to consider synonyms as well?

```
======================================================= short test summary info =======================================================
FAILED test_ARAX_overlay.py::test_predict_drug_treats_disease_virtual - AssertionError: assert 'probably_treats' in Counter({'contra...
FAILED test_ARAX_overlay.py::test_predict_drug_treats_disease_attribute - assert 0 > 0
FAILED test_ARAX_overlay.py::test_issue_832 - AssertionError: assert 'probably_treats' in Counter({'contraindicated_for': 330, 'indi...
```
```
============================================================== FAILURES ===============================================================
______________________________________________ test_predict_drug_treats_disease_virtual _______________________________________________

    def test_predict_drug_treats_disease_virtual():
        query = {"previous_message_processing_plan": {"processing_actions": [
            "create_message",
            "add_qnode(curie=DOID:1588, id=n0, type=disease)",
            "add_qnode(type=chemical_substance, id=n1)",
            "add_qedge(source_id=n0, target_id=n1, id=e0)",
            "expand(edge_id=e0)",
            "overlay(action=predict_drug_treats_disease, source_qnode_id=n1, target_qnode_id=n0, virtual_relation_label=P1)",
            "resultify()",
            "return(message=true, store=false)",
        ]}}
        [response, message] = _do_arax_query(query)
        print(response.show())
        assert response.status == 'OK'
>       _virtual_tester(message, 'probably_treats', 'P1', 'probability_treats', 'data:0951', 2)

test_ARAX_overlay.py:397: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = {'_datetime': '2020-07-04 13:37:53',
 'code_description': 'Normal completion',
 'context': 'https://raw.githubusercont...ce', 'essence', 'essence_type'],
 'terms': None,
 'tool_version': 'ARAX 0.6.0',
 'type': 'translator_reasoner_message'}
edge_type = 'probably_treats', relation = 'P1', attribute_name = 'probability_treats', attribute_type = 'data:0951'
num_different_values = 2

    def _virtual_tester(message: Message, edge_type: str, relation: str, attribute_name: str, attribute_type: str, num_different_values=2):
        """
        Tests overlay functions that add virtual edges
        message: returned from _do_arax_query
        edge_type: the name of the virtual edge (eg. has_jaccard_index_with)
        relation: the relation you picked for the virtual_edge_relation (eg. N1)
        attribute_name: the attribute name to test (eg. 'jaccard_index')
        attribute_type: the attribute type (eg. 'data:1234')
        num_different_values: the number of distinct values you wish to see have been added as attributes
        """
        edge_types_in_kg = Counter([x.type for x in message.knowledge_graph.edges])
>       assert edge_type in edge_types_in_kg
E       AssertionError: assert 'probably_treats' in Counter({'contraindicated_for': 165, 'indicated_for': 1})

test_ARAX_overlay.py:75: AssertionError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Response:
  status: OK
  n_errors: 0  n_warnings: 0  n_messages: 64

_____________________________________________ test_predict_drug_treats_disease_attribute ______________________________________________

    def test_predict_drug_treats_disease_attribute():
        query = {"previous_message_processing_plan": {"processing_actions": [
            "create_message",
            "add_qnode(curie=DOID:1588, id=n0)",
            "add_qnode(type=chemical_substance, id=n1)",
            "add_qedge(source_id=n0, target_id=n1, id=e0)",
            "expand(edge_id=e0)",
            "overlay(action=predict_drug_treats_disease)",
            "resultify()",
            "return(message=true, store=false)",
        ]}}
        [response, message] = _do_arax_query(query)
        print(response.show())
        assert response.status == 'OK'
>       _attribute_tester(message, 'probability_treats', 'data:0951', 2)

test_ARAX_overlay.py:414: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = {'_datetime': '2020-07-04 13:38:11',
 'code_description': 'Normal completion',
 'context': 'https://raw.githubusercont...ce', 'essence', 'essence_type'],
 'terms': None,
 'tool_version': 'ARAX 0.6.0',
 'type': 'translator_reasoner_message'}
attribute_name = 'probability_treats', attribute_type = 'data:0951', num_different_values = 2

    def _attribute_tester(message, attribute_name: str, attribute_type: str, num_different_values=2):
        """
        Tests attributes of a message
        message: returned from _do_arax_query
        attribute_name: the attribute name to test (eg. 'jaccard_index')
        attribute_type: the attribute type (eg. 'data:1234')
        num_different_values: the number of distinct values you wish to see have been added as attributes
        """
        edges_of_interest = []
        values = set()
        for edge in message.knowledge_graph.edges:
            if hasattr(edge, 'edge_attributes'):
                for attr in edge.edge_attributes:
                    if attr.name == attribute_name:
                        edges_of_interest.append(edge)
                        assert attr.type == attribute_type
                        values.add(attr.value)
>       assert len(edges_of_interest) > 0
E       assert 0 > 0
E        +  where 0 = len([])

test_ARAX_overlay.py:60: AssertionError
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Response:
  status: OK
  n_errors: 0  n_warnings: 0  n_messages: 65

___________________________________________________________ test_issue_832 ____________________________________________________________

    def test_issue_832():
        query = {"previous_message_processing_plan": {"processing_actions": [
            "create_message",
            "add_qnode(curie=DOID:1588, id=n0)",
            "add_qnode(type=chemical_substance, id=n1)",
            "add_qedge(source_id=n0, target_id=n1, id=e0)",
            "expand(edge_id=e0)",
            "overlay(action=predict_drug_treats_disease, source_qnode_id=n1, target_qnode_id=n0, virtual_relation_label=P1)",
            "resultify()",
            "return(message=true, store=false)",
        ]}}
        [response, message] = _do_arax_query(query)
        print(response.show())
        assert response.status == 'OK'
>       _virtual_tester(message, 'probably_treats', 'P1', 'probability_treats', 'data:0951', 2)

test_ARAX_overlay.py:431: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = {'_datetime': '2020-07-04 13:38:30',
 'code_description': 'Normal completion',
 'context': 'https://raw.githubusercont...ce', 'essence', 'essence_type'],
 'terms': None,
 'tool_version': 'ARAX 0.6.0',
 'type': 'translator_reasoner_message'}
edge_type = 'probably_treats', relation = 'P1', attribute_name = 'probability_treats', attribute_type = 'data:0951'
num_different_values = 2

    def _virtual_tester(message: Message, edge_type: str, relation: str, attribute_name: str, attribute_type: str, num_different_values=2):
        """
        Tests overlay functions that add virtual edges
        message: returned from _do_arax_query
        edge_type: the name of the virtual edge (eg. has_jaccard_index_with)
        relation: the relation you picked for the virtual_edge_relation (eg. N1)
        attribute_name: the attribute name to test (eg. 'jaccard_index')
        attribute_type: the attribute type (eg. 'data:1234')
        num_different_values: the number of distinct values you wish to see have been added as attributes
        """
        edge_types_in_kg = Counter([x.type for x in message.knowledge_graph.edges])
>       assert edge_type in edge_types_in_kg
E       AssertionError: assert 'probably_treats' in Counter({'contraindicated_for': 330, 'indicated_for': 2})

test_ARAX_overlay.py:75: AssertionError

```

Note: In order to run DSL in this branch, you'll need to build/get the `NodeSynonymizer`'s sqlite database following the steps Eric described [here](https://github.com/RTXteam/RTX/issues/825#issuecomment-653652955).

**Comment by dkoslicki** [comment]
@amykglen Ok, so there are a few things going on here:
1. The mapping file for the ML method uses chembl, HP, and DOID. CURIES, so to update this, we will either need to:
    1. Map the existing chembl, hp, DOID etc. nodes to the "preferred" ones (this will need to be re-done any time what counts as the "preferred" curie changes. 
    1. Leave somewhere in the node objects a/the list of equivalent curies so I can loop over them and find ones that the ML method wants as input (I like this option)
    1. Maybe there's some way to use the `node_synonymizer.py` to do this mapping that I desire? I'm not familiar with that yet since it's so new, but if you can let me know if this would work for my "map back to chembl/doid" purposes, that would be great.
2. It appears that while the `id` on the nodes have been updated, the `uri` is still pointing to the chembl. eg:
```
{'description': 'None',
  'id': 'CHEBI:136051',
  'name': 'lusutrombopag',
  'node_attributes': [],
  'qnode_ids': ['n1'],
  'symbol': None,
  'type': ['chemical_substance'],
  'uri': 'https://www.ebi.ac.uk/chembl/compound/inspect/CHEMBL2107831'}]
```

I assume that this `uri` field is in error (since otherwise I could just parse that to pull off the original CHEMBL, DOID, HP, whatever), correct?

**Comment by amykglen** [comment]
@dkoslicki - re: ii) and iii) - you can use `node_synonymizer.py` to get the mappings you want, but I'm also happy to just stick the equivalent curies in a `NodeAttribute` if desired. not sure if others are very opinionated on this... but maybe at least to start we could go with the NodeAttribute method?

re: 2. yeah, that wouldn't be a good field to rely on (this gets at an issue we'll have to deal with of 'merging' the properties on nodes...)

**Comment by dkoslicki** [comment]
@amykglen I'm agnostic about including it in the `NodeAttributes` or using the `node_synonymizer.py.` But since I've been able to figure out how the `node_synonymizer.py` works now, how about I just give that a crack and see if I can fix it that way?

**Comment by amykglen** [comment]
ok, sounds good!

**Comment by dkoslicki** [comment]
Ok, so since the synonymizer can return multiple curies that are in the ML model, I've decided to loop over all of them and just take whichever pair of equivalent curies (that are in the ML model) give the highest probability.

Note that I am using `synonymizer.get_normalizer_results(input_curie, kg_name='ARAX')` so @edeutsch I don't know if this is correct, but I was getting the following error when using the SRI:
```
File "/home/dkoslicki/Desktop/RTX/code/ARAX/ARAXQuery/Expand/../../NodeSynonymizer/node_synonymizer.py", line 1177, in get_normalizer_results
    results[entity] = result[entity]
TypeError: 'NoneType' object is not subscriptable
```
which did not occur when using `kg_name='ARAX'`

**Comment by edeutsch** [comment]
kg_name should be either KG1 or KG2 or SRI
ARAX is not a permitted value, but it shouldn't crash either. will fix.

**Comment by dkoslicki** [comment]
@edeutsch Ah, I was looking at the `help()` on the class and it only mentioned SRI and ARAX, so I didn't think to look at the code itself...

**Comment by dkoslicki** [comment]
Ok, so all tests are passing now, but the predictions are different:
Old system:
```
CHEMBL.COMPOUND:CHEMBL1144 contraindicated_for DOID:1588
Provenance: MyChem.info
relation: contraindicated_for
is_defined_by: ARAX/KG1
id: CHEMBL.COMPOUND:CHEMBL1144--contraindicated_for--DOID:1588
probability_treats (data:0951): 0.342464008983882
```
Now with node synonymization:
```
CHEBI:63618 contraindicated_for MONDO:0002049
Provenance: MyChem.info
relation: contraindicated_for
is_defined_by: ARAX/KG1
id: KG1:3030164
probability_treats (data:0951): 0.6754644125431072
```
So with the same drug and disease pair, the prediction went up. This is probably due to my taking the max prediction based on curie synonym, but imho, this makes the most sense since the curies _should_ be synonymous...
In any case, @amykglen, check to see if it's kosher on your end as well, and we can close this.

Note @chunyuma for #758, we will need to change [these lines](https://github.com/RTXteam/RTX/blob/node-normalization/code/ARAX/ARAXQuery/Overlay/predict_drug_treats_disease.py#L41-L52) to then point to KG2, as well as remove the hack when we weren't using the correct CURIE prefixes

**Comment by amykglen** [comment]
awesome! yep, looks great from my end - all `predict_drug_treats_disease` tests are now passing...

(we're almost at a point where `node-normalization` is ready to merge into master - just a couple failing tests remaining, which I'll write up in separate issues...)

**Comment by edeutsch** [comment]
I made several more fixes to node_normalizer.py. Please git pull from master and check to make sure this is better. Was crashing on unrecognized curies.

**Comment by amykglen** [comment]
sounds good @edeutsch - I merged the changes into the `node-normalization` branch and adjusted as necessary (for the change in output format re: unrecognized curies).

things are still running smoothly with the updates.

**Comment by dkoslicki** [comment]
fixed with 7fc97b99d2902cec94a3b0958d795dff2e9951b1 closing
